drivers/soundwire/generic_bandwidth_allocation.c

```diff
@@ -372,16 +410,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 	}
 
 	if (i == clk_values) {
-		dev_err(bus->dev, "%s: could not find clock value for bandwidth %d\n",
+multilane:
+		dev_dbg(bus->dev, "%s: could not find clock value for bandwidth %d, checking multi-lane\n",
 			__func__, bus->params.bandwidth);
-		return -EINVAL;
+
+		m_rt = list_last_entry(&bus->m_rt_list, struct sdw_master_runtime, bus_node);
+		/*
+		 * Find available manager lanes that connected to the first Peripheral. No need
+		 * to check all Peripherals available lanes because we can't use multi-lane
+		 * if we can't find any available lane for the first Peripheral.
+		 */
+		s_rt = list_first_entry(&m_rt->slave_rt_list, struct sdw_slave_runtime, m_rt_node);
+		slave_prop = &s_rt->slave->prop;
+
+		for (i = 1; i < SDW_MAX_LANES; i++) {
+			if (!slave_prop->lane_maps[i])
+				continue;
+
+			required_bandwidth = 0;
+			list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
+				required_bandwidth += m_rt->stream->params.rate *
+					hweight32(m_p_rt->ch_mask) *
+					m_rt->stream->params.bps;
+			}
+			if (required_bandwidth <= curr_dr_freq - bus->lane_used_bandwidth[i]) {
+				/* Check if m_lane is connected to all Peripherals */
+				if (!check_all_peripherals_connected(m_rt, slave_prop->lane_maps[i])) {
+					dev_dbg(bus->dev,
+						"some Peripherals are not connected to %d\n",
+						slave_prop->lane_maps[i]);
+					continue;
+				}
+				m_lane = slave_prop->lane_maps[i];
+				dev_dbg(&s_rt->slave->dev,
+					"M lane %d P lane %d can be used\n",
+					m_lane, i);
```

Comment:
without much of a background on multi-lane, it is very difficult to understand what M lane and P lane is.

Reply:
> without much of a background on multi-lane, it is very difficult to understand what M lane and P lane is.

A Manager can support multiple lanes, like lane 0, lane 1, and lane 2, lane 3. And a Peripheral also can have multiple lanes. Manager lane 0 and Peripheral lane 0 should be always connected. However, other lanes can be connected per request and no need to be 1:1 mapping. For example, Manager lane 1 can connect to Peripheral lane 2. So that we use M lane to represent Manager lane and P lane to Peripheral lane.